### PR TITLE
Miles panic - save registers, panic_exception_entry

### PR DESCRIPTION
**Pull request by milesmaltbie**

### CODE REVIEW

File: tos/platforms/dev6a/startup.c [hunk 1]
```diff
@@ -127,7 +127,7 @@ extern void owl_strange2gold(uint32_t loc);
 extern void owl_startup();
 
 /* see tos/system/panic/PanicP.nc */
-extern void __panic_exception_entry(uint32_t exception);
+extern void __panic_exception_entry(uint32_t exception, uint32_t exc_psr, uint32_t exc_msp);
```
Comment (cire831) [inline tos/platforms/dev6a/startup.c]: nice.  right track.

we'll see how the code plays out in the PM.  We need to see what kind of machine instructions this puppy generates.

File: tos/platforms/dev6a/startup.c [hunk 2]
```diff
@@ -168,9 +168,7 @@ void handler_debug(uint32_t exception) {
   t0 = USECS_VAL;
   while ((USECS_VAL - t0) < WIGGLE_DELAY) ;
 
-  ROM_DEBUG_BREAK(0);
-
-#ifdef HANDLER_FAULT_WAIT
+#ifdef HANDLER_FAULT_WAIT_Save
```
Comment (cire831) [inline tos/platforms/dev6a/startup.c]: why did you change the name?

If you have a good reason, great.   ...

File: tos/platforms/dev6a/startup.c [hunk 3]
```diff
@@ -198,10 +196,14 @@ void HardFault_Handler() {
 void __default_handler()  __attribute__((interrupt));
 void __default_handler()  {
   uint32_t exception;
+  uint32_t exc_psr;
+  uint32_t exc_msp;
 
-  exception = __get_xPSR() & 0x1ff;
+  exc_psr = __get_xPSR();
+  exception = exc_psr & 0x1ff;
+  exc_msp = __get_MSP();
```
Comment (cire831) [inline tos/platforms/dev6a/startup.c]: not bad.  do these generate inline or do they actually create a procedure call?

File: tos/platforms/mm6a/startup.c [hunk 4]
```diff
@@ -113,10 +113,10 @@ const image_info_t image_info __attribute__ ((section(".image_meta"))) = {
 ow_control_block_t ow_control_block __attribute__ ((section(".overwatch_data")));
 
 /* space to save registers during crash processing */
-uint32_t crash_regs[CRASH_REGS_BYTES]   __attribute__ ((section(".crash_regs")));
+uint32_t crash_regs[CRASH_REGS_WORDS]   __attribute__ ((section(".crash_regs")));
```
Comment (cire831) [inline tos/platforms/mm6a/startup.c]: i thought we decided to get rid of crash_regs and just slam everything onto the crash_stack similar to how CrashCatcher_armv7m.S does it.

Did I miss something?

okay.  I see that you don't use crash_regs.  I got confused because you modified the above line.  Gratuitous.  no biggie

Since we agreed to remove crash_regs, it would have been much better to remove all references to crash_regs.

File: tos/system/panic/PanicP.nc [hunk 5]
```diff
@@ -388,13 +368,49 @@ implementation {
...
+      pac->bxRegs[12] = exc_regs->r12;
+    } else {
+
+      i = 0;
+      while (i < 4) {
+        pac->bxRegs[i] = (uint32_t)(old_sp + i);
+        i++;
+      }
+      pac->bxLR = crash_stack[9 + 118];
+      pac->bxRegs[12] = crash_stack[126];
```
Comment (cire831) [inline tos/system/panic/PanicP.nc]: the above two lines are extremely prone to error.

we need to figure out a different way to do this.  We'll figure it out tomarrow unless you beat me to it.

I realize this is a work in progress and the above is quite likely to just be a quick and dirty to get something working as the vehicle for further work.

A better way is to lay a structure on top of the raw memory.

File: tos/platforms/mm6a/startup.c [hunk 6]
```diff
@@ -113,10 +113,10 @@ const image_info_t image_info __attribute__ ((section(".image_meta"))) = {
 ow_control_block_t ow_control_block __attribute__ ((section(".overwatch_data")));
 
 /* space to save registers during crash processing */
-uint32_t crash_regs[CRASH_REGS_BYTES]   __attribute__ ((section(".crash_regs")));
```
Comment (cire831) [inline tos/platforms/mm6a/startup.c]: I noticed that dev6a/startup.c changed CRASH_{REG,STACK}_BYTES to CRASH_{REG,STACK}_WORDS in an earlier commit.

It is messy and leads to confusion splitting the same change across multiple commits.

that is as much a ding on me as on us.  Part of the problem is essentially duplicated code.  I have also been guilty of splitting across commits.  I try not to.

It should be avoided and we should try to be more consistent.

File: tos/platforms/mm6a/startup.c [hunk 7]
```diff
@@ -373,26 +375,12 @@ void (* const __vectors[])(void) __attribute__ ((section (".vectors"))) = {
...
-  P7MAP->PMAP_REG[4] = PMAP_UCB0CLK;
-  P7MAP->PMAP_REG[5] = PMAP_UCB0SIMO;
-  P7MAP->PMAP_REG[7] = PMAP_UCA2CLK;
-
+  PMAP->KEYID = PMAP_KEYID_VAL;
+  P2MAP->PMAP_REG[5] = PMAP_SMCLK;
+  P7MAP->PMAP_REG[0] = PMAP_UCA1CLK;
+  P7MAP->PMAP_REG[1] = PMAP_UCA1SOMI;
+  P7MAP->PMAP_REG[2] = PMAP_UCA1SIMO;
+  P7MAP->PMAP_REG[3] = PMAP_TA1CCR1A;
```
Comment (cire831) [inline tos/platforms/mm6a/startup.c]: ah.  this is really bad.

any idea of how this happened?

you changed the port mapping on the mm6a to the mapping used on the dev6a.

just won't work and is very wrong.

I don't want you to be defensive.  I would like you to think about how this happened so it doesn't happen again.

File: tos/platforms/mm6a/startup.c [hunk 8]
```diff
@@ -425,28 +413,90 @@ void __watchdog_init() {
...
-  P4->OUT = 0x30; P4->DIR = 0xFD;
-  P5->OUT = 0x81; P5->DIR = 0xA7;
-  P6->OUT = 0x18; P6->DIR = 0x18;
-  P7->OUT = 0xF9; P7->DIR = 0xF8;
-  P8->OUT = 0x00; P8->DIR = 0x02;
-  PJ->OUT = 0x04; PJ->DIR = 0x06;
 
   /*
-   * gps_cts has a pull up so that the gps comes up in UART mode.
+   * see hardware.h for initial values and changed mappings
```
Comment (cire831) [inline tos/platforms/mm6a/startup.c]: this is also really bad.

running dev6 port defs and port_mapping on the mm6 is a very bad thing (tm)

DO NOT run this code on the mm6.

I do not know how it will behave.  It could cause bad things to happen.

File: tos/system/panic/PanicP.nc [hunk 9]
```diff
@@ -388,13 +368,49 @@ implementation {
     pap->a1    = old_sp[3];
     pap->a2    = old_sp[4];
     pap->a3    = old_sp[5];
-    debug_break(1);
-    if (pcb.in_panic) {
-      pcb.in_panic |= 0x80;             /* flag a double */
-      ROM_DEBUG_BREAK(0xf1);
-      call OverWatch.strange(0x83);     /* no return */
+
+    pac = &b0p->crash_info;
```
Comment (cire831) [inline tos/system/panic/PanicP.nc]: i thought we were going to change pac to something more meaningful like cip (crash_info pointer)?

File: tos/platforms/mm6a/startup.c (hunk 7 above)
Comment (milesmaltbie) [reply tos/platforms/mm6a/startup.c]: This was due to a mistake that happened quickly right before I committed. I didn't notice it until after the commit when you pointed it out to me, and I have a quick and easy way to get it back to normal.

I will think about how this happened so that it doesn't happen again.

File: tos/platforms/mm6a/startup.c (hunk 8 above)
Comment (milesmaltbie) [reply tos/platforms/mm6a/startup.c]: This is also due to the mistake above. It was an error that I can fix and will make sure doesn't happen again.

File: tos/platforms/mm6a/startup.c (hunk 4 above)
Comment (milesmaltbie) [reply tos/platforms/mm6a/startup.c]: I will remove all references to crash_regs before next commit.

File: tos/platforms/dev6a/startup.c (hunk 3 above)
Comment (milesmaltbie) [reply tos/platforms/dev6a/startup.c]: They seem to generate inline, as I noticed the register values don't change when they execute. I can look into that further.

File: tos/platforms/dev6a/startup.c (hunk 2 above)
Comment (milesmaltbie) [reply tos/platforms/dev6a/startup.c]: I changed the name to get past the HANDER_FAULT_WAIT so I could work on panic_exception_entry. I was planning on changing the name back after testing and would like to know more about how HANDER_FAULT_WAIT works.

File: tos/system/panic/PanicP.nc (hunk 5 above)
Comment (milesmaltbie) [reply tos/system/panic/PanicP.nc]: I'll look at these two lines and see if I can come up with something better before we meet. 

One thing I would like to do is create constant values that we can use instead of number calculations.

File: tos/system/panic/PanicP.nc (hunk 9 above)
Comment (milesmaltbie) [reply tos/system/panic/PanicP.nc]: I'll change pac to cip (crash_info_pointer). I was waiting to discuss it with you, but I have no issues with cip and it seems to be good enough to just go with that.

I'll make these changes before we meet today.

File: tos/platforms/mm6a/startup.c (hunk 6 above)
Comment (milesmaltbie) [reply tos/platforms/mm6a/startup.c]: Would you prefer that I keep the name CRASH_STACK_BYTES and change the numerical value to reflect word access? 

The reason for this change is to make it easier to access the registers as words later on.

File: tos/platforms/dev6a/startup.c (hunk 1 above)
Comment (milesmaltbie) [reply tos/platforms/dev6a/startup.c]: Looking forward to testing out the code together.
